pkgs/top-level/default.nix

```diff
@@ -49,16 +68,42 @@ let # Rename the function arguments
 in let
   lib = import ../../lib;
 
-  inherit (lib) throwIfNot;
+  inherit (lib) intersectAttrs throwIfNot throwIf;
+
+  legacySystemArgsUsed = intersectAttrs { system = null; localSystem = null; crossSystem = null; } args;
+
+  intentToCrossCompile =
+    # is it legacy cross?
+    (args?localSystem && !lib.systems.equals crossSystem localSystem)
+    || # is it new cross?
+    (args?buildPlatform && args?hostPlatform && !lib.systems.equals crossSystem localSystem);
 
   checked =
     throwIfNot (lib.isList overlays) "The overlays argument to nixpkgs must be a list."
     lib.foldr (x: throwIfNot (lib.isFunction x) "All overlays passed to nixpkgs must be functions.") (r: r) overlays
     throwIfNot (lib.isList crossOverlays) "The crossOverlays argument to nixpkgs must be a list."
     lib.foldr (x: throwIfNot (lib.isFunction x) "All crossOverlays passed to nixpkgs must be functions.") (r: r) crossOverlays
+    throwIf (args?buildPlatform && !args?hostPlatform) ''
+      nixpkgs: You have only specified buildPlatform, but not hostPlatform.
+        - For cross compilation: specify both
+        - For native compilation: just specify hostPlatform; buildPlatform defaults to it
+    ''
+    throwIf (args?hostPlatform && (legacySystemArgsUsed != {})) ''
+      nixpkgs: You have specified hostPlatform, but also one of the system, localSystem, or crossSystem.
+      Mixing legacy system arguments and hostPlatform is not supported.
+      ${if intentToCrossCompile then ''
+        For cross compilation, please specify the nixpkgs arguments
+          - hostPlatform: the platform on which the packages will run
+          - buildPlatform: the platform on which the derivations are built
+      '' else ""}
+      If you are not cross-compiling, please specify only hostPlatform.
```

Comment:
This conflicts with like 89 above where it implies you can set both `hostPlatform` and `buildPlatform` it's just that `buildPlatform` = `hostPlatform` by default so you don't need to set it.